bgpd/bgp_zebra.c

```diff
@@ -1577,6 +1577,37 @@ void bgp_zebra_withdraw(const struct prefix *p, struct bgp_path_info *info,
 	zclient_route_send(ZEBRA_ROUTE_DELETE, zclient, &api);
 }
 
+/* Withdraw all entries in a BGP instances RIB table from Zebra */
+void bgp_zebra_withdraw_table(struct bgp *bgp, afi_t afi, safi_t safi)
+{
+	struct bgp_dest *dest;
+	struct bgp_table *table;
+	struct bgp_path_info *pi;
+
+	/* Don't try to withdraw table if we're not connected to Zebra or Zebra
+	   doesn't know of this instance. */
+	if (!bgp_install_info_to_zebra(bgp))
+		return;
+
+	/* Select and check RIB table for given afi/safi */
+	table = bgp->rib[afi][safi];
+	if (!table)
+		return;
+
+	/* Iterate over RIB and withdraw routes from Zebra */
+	for (dest = bgp_table_top(table); dest; dest = bgp_route_next(dest)) {
+		for (pi = bgp_dest_get_bgp_path_info(dest); pi; pi = pi->next) {
+			if (CHECK_FLAG(pi->flags, BGP_PATH_SELECTED)
+			    && (pi->type == ZEBRA_ROUTE_BGP
+				&& (pi->sub_type == BGP_ROUTE_NORMAL
```

Comment:
Do we care about aggregated networks here? I meant `BGP_ROUTE_AGGREGATE`.

Reply:
I used exactly the same selection process already used to announce routes, so everything previously announced should get withdrawn. Did i miss some special aggregate route announcement procedure?

Reply:
I mean if I create an aggregate network it creates blackhole network in Zebra. Are they withdrawn as well?